font weight selector

danilkhromov opened the pull request:
This is a proposed implementation for font weight selection that adds following features:

  * font weight setting for editor font
  * font weight selector for editor font

This closes #3161 

This is how font weight selector looks like in settings:
![font weight selector](https://user-images.githubusercontent.com/26026560/111542355-e3edd180-8771-11eb-9f59-713735751436.png)

And here is font weight change in action:
![font weight change](https://user-images.githubusercontent.com/26026560/111542388-f405b100-8771-11eb-8523-bf08e41d7b38.gif)

Test on my machine shows that if selected font weight does not exists then closest to it will be rendered.

Commit:
font weight selector

danilkhromov committed:
font weight selector

danilkhromov committed:
Merge branch 'feature/font-weight-selector' of github.com:danilkhromov/insomnia into feature/font-weight-selector

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

develohpanda commented on packages/insomnia-app/app/common/constants.js:
```diff
@@ -182,6 +182,17 @@ export const MIN_INTERFACE_FONT_SIZE = 8;
+  { label: 'Bold', value: 700 },
+  { label: 'ExtraBold', value: 800 },
+  { label: 'Heavy', value: 900 },
+];
```
Using an array for this purpose works but it seems better suited as an object/lookup 🤔 

```js
export const EDITOR_FONT_WEIGHTS = {
  Thin: 100,
  UltraLight: 200,
  Light: 300,
  ...
}
```

develohpanda commented on packages/insomnia-app/app/models/settings.js:
```diff
@@ -96,6 +97,7 @@ export function init(): BaseSettings {
     clearOAuth2SessionOnRestart: true,
     fontInterface: null,
     fontMonospace: null,
+    fontMonospaceWeight: 400,
```
After converting `EDITOR_FONT_WEITGHTS` to an object, this should be set using `EDITOR_FONT_WEITGHTS.Normal` instead of hard-coded. That way the number doesn't seem magical to the next person looking at it 😄

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -108,6 +109,11 @@ class General extends React.PureComponent<Props, State> {
+  async _handleFontWeightChange(el: SyntheticEvent<HTMLInputElement>) {
+    const settings = await this._handleUpdateSetting(el);
+    setFont(settings);
+  }
```
This seems to be a duplicate of both `_handleFontSizeChange` and `_handleFontChange`. Maybe all the font related items (font, size, weight) could just use the `_handleFontChange` handler?

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -326,10 +332,34 @@ class General extends React.PureComponent<Props, State> {
+                  name="fontMonospaceWeight"
+                  value={settings.fontMonospaceWeight}
+                  onChange={this._handleFontWeightChange}>
+                  {EDITOR_FONT_WEITGHTS.map((item, index) => (
+                    <option key={index} value={item.value}>
+                      {item.label}
+                    </option>
+                  ))}
```
It's good to avoid using array index as keys in React. While it's not as concerning in this particular use-case, it is still an [anti-pattern](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318).

After converting `EDITOR_FONT_WEITGHTS` to an object, you can use `Object.entries(EDITOR_FONT_WEITGHTS)` to get the key-value pairs.

```suggestion
                  {Object.entries(EDITOR_FONT_WEITGHTS).map(e => (
                    <option key={e[1]} value={e[1]}>
                      {e[0]}
                    </option>
                  ))}
```

develohpanda reviewed (changes requested):
Thank you for the PR, and welcome to the project!

This seems like a useful addition. I haven't had a chance to test this locally yet but I just had a couple of minor observations in the code that would be worth addressing 🙂 

Also, since this contains some UI changes, could you please add a screenshot or gif into the PR description? Thanks!

danilkhromov committed:
use object for font weight seletion

convert EDITOR_FONT_WEGIHTS from array to objects
handle all font related changes in one method

danilkhromov commented:
Thanks for taking your time to review my pr, and providing useful suggestions

As suggested, i improved following things:
* Converted `EDITOR_FONT_WEITGHTS` from array to object
* Now all font changes handled in one handler
* Added some screenshots/gif to the description

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

develohpanda reviewed (approved):
Looks good! Seems to work as well on macOS.

I'm not so familiar with this, but is there any way to determine which weights of a font are available, so we only show those weights instead of all?

Pinging @nijikokun for product review

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov commented:
@develohpanda i was not yet able to find an elegant solution that will allow display only installed weights for each font. But i can give it another try this weekend maybe.

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
show only installed weights for selected font

danilkhromov commented:
@develohpanda so i think i found a solution that works and also not very hard to understand.

now only installed weights are shown for selected font:
![show only installed weights](https://user-images.githubusercontent.com/26026560/111906052-138d2a00-8a4f-11eb-9334-54350b2bb889.gif)

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

nijikokun commented:
Great work @danilkhromov looks good.

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -71,10 +83,22 @@ class General extends React.PureComponent<Props, State> {
+    const fontsMono = allFonts
+      .filter(i => i.monospace || i.family.match(FORCED_MONO_FONT_REGEX && !i.italic))
+      .map(({ family, monospace }) => {
+        const weights = allFonts
+          .filter(i => i.family === family)
+          .map(item => item.weight)
+          .filter((item, index, self) => self.indexOf(item) === index)
+          .sort((a, b) => (a > b ? 1 : -1));
```
This logic to extract the font weights for a particular font might be better suited to a traditional `forEach` and an ES6 `Set` to achieve the same result in fewer iterations.

```js
import { ascendingNumberSort } from '../../../common/sorting';
...
function getWeightsForFont(font: string): Array<number> {
  const weightSet = new Set();
  
  allFonts.forEach(({family, weight}) => {
    if (family === font)
      weightSet.add(weight);
    }
  });

  return [...weightSet].sort(ascendingNumberSort);
}
```

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -71,10 +83,22 @@ class General extends React.PureComponent<Props, State> {
-    const fontsMono = fonts.filter(i => i.monospace || i.family.match(FORCED_MONO_FONT_REGEX));
+    const fontsMono = allFonts
+      .filter(i => i.monospace || i.family.match(FORCED_MONO_FONT_REGEX && !i.italic))
+      .map(({ family, monospace }) => {
```
Weights are extracted for all monospaced fonts in the `componentDidMount` event, even if we only use one at a time, so there is some scope for improvement here given a user may have many fonts installed and the current implementation has several nested loops.

It think it might be more suitable to extract and show the font-weights only for the font that is selected. You could do this by storing a `editorFontWeights: {[string]: string}` key-value object in the state, separate to `fontsMono`, and change that `editorFontWeights` object as the selected font changes. This way we are only extracting data that's needed, when it's needed, and nothing more.

danilkhromov replied:
Thanks again for reviewing my pr and sorry for taking so long to implement requested changes. So now only weights for selected fonts are extracted and they are changed when selected font changes. I only declared `editorFontWeights` as an `Array<number>` since weight names can differ from font to font so its safer to use number in this case

develohpanda replied:
No worries! Thank you for the updates. I have it on my list to review again 🙂

develohpanda reviewed (changes requested):
Nice work figuring out how to filter the font weights! As a user it certainly looks more polished now 🤗

After the changes for filtering fonts I've noticed some improvements that are worth making so marking this PR as such. Let me know how you get on, happy to help get it across the line! 🙂

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
find weights only for selected fonts

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

danilkhromov committed:
Merge branch 'develop' into feature/font-weight-selector

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -71,14 +86,21 @@ class General extends React.PureComponent<Props, State> {
       .sort((a, b) => (a.family > b.family ? 1 : -1));
 
-    // Find monospaced fonts
+    // Find monospaced fonts with installed weights for each font family
```
Good on you for keeping the comments up to date, very much appreciated 🙌🏽

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -71,14 +86,21 @@ class General extends React.PureComponent<Props, State> {
     //  - https://github.com/Kong/insomnia/issues/1835
-    const fontsMono = fonts.filter(i => i.monospace || i.family.match(FORCED_MONO_FONT_REGEX));
+    const fontsMono = allFonts
+      .filter(i => i.monospace || i.family.match(FORCED_MONO_FONT_REGEX && !i.italic))
```
Are you sure that the `!i.italic` check is in the right place? It seems like it's accidentally part of the `.match()` call. Let me know if that's not the case, I may be missing something 🙂

danilkhromov replied:
my bad
fixed now :smiley:

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -326,10 +371,36 @@ class General extends React.PureComponent<Props, State> {
+                  name="fontMonospaceWeight"
+                  value={settings.fontMonospaceWeight}
+                  onChange={this._handleFontChange}>
+                  {editorFontWeights.map((item, index) => (
+                    <option key={index} value={item}>
```
Definitely want to [avoid using index as key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). `item` here should always be unique, so we can key with that.

```suggestion
                  {editorFontWeights.map(item => (
                    <option key={item} value={item}>
```

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -108,14 +130,37 @@ class General extends React.PureComponent<Props, State> {
   }
 
   async _handleFontChange(el: SyntheticEvent<HTMLInputElement>) {
+    const currentTarget = { name: el.currentTarget.name, value: el.currentTarget.value };
```
Prime use case for destructuring! 😄 

```suggestion
    const { currentTarget } = el;
```

danilkhromov replied:
for some reason if i do it like that then actual `value` that im getting to find weights is previously selected font instead of the new one :thinking: 
do you know why it behaves like that?

develohpanda replied:
That's...strange, not sure off the top of my head, but if it doesn't work keep it like you had it. My bad! 😔

develohpanda commented on packages/insomnia-app/app/ui/components/settings/general.js:
```diff
@@ -108,14 +130,37 @@ class General extends React.PureComponent<Props, State> {
+      const weightSet = new Set();
 
-    setFont(settings);
+      const fonts = fontScanner.findFontsSync({ family: fontFamily });
```
This seems OK to me, but if calling into `fontScanner` too frequently does become an issue we may have to explore a different route in future. No changes required right now, though, I think it'll be fine!

Also, I wonder if the issue I noticed on Windows is related to this... 🤔 `fontScanner` has caused tricky issues in the past

develohpanda reviewed (commented):
Code generally looks good to me, but while testing on Windows it doesn't seem to be working as intended. 

![](http://g.recordit.co/jjOCt574U9.gif)

I will be away for the next week or so, and can take a swing at Windows when I'm back, in case you don't have access to a Windows dev machine. Also, don't worry about keeping up to date with `develop`, I'll make sure it's up to date before merging. 🤗 Thank you for your patience!

danilkhromov commented:
@develohpanda i have a windows in the vm, so hopefully i will be able to reproduce this behavior there. Im also assuming that this is caused by the `fontScanner`, seems like it returns slightly different data on windows. 
I will look into that on the weekend and also will take care of other things that you mentioned (sorry if there is some obvious mistakes, im mostly developing backend services and i dont have very deep knowledge of js and react). 

Hopefully i will be able to tackle the windows issue, but if not, then your help will be much appreciated :slightly_smiling_face:

develohpanda commented:
Open-source is a really great way to learn and practice, so no worries at all! and no rush, take your time 🤗

danilkhromov commented:
> Code generally looks good to me, but while testing on Windows it doesn't seem to be working as intended.
> 
> ![](https://camo.githubusercontent.com/a8e78c7d6e9aeaefb3faf654f5d6dfc85f99dd4e8e1c023d8e1eeb5a4a9f8d33/687474703a2f2f672e7265636f726469742e636f2f6a6a4f437435373455392e676966)
> 
> I will be away for the next week or so, and can take a swing at Windows when I'm back, in case you don't have access to a Windows dev machine. Also, don't worry about keeping up to date with `develop`, I'll make sure it's up to date before merging. hugs Thank you for your patience!

after some investigation it looks like `fontScanner` on windows returns different data compared to what i have on linux (and i assume on macOS as well). `family` in the font object on windows contains full name of each installed variant e.g. `JetBrains Mono Medium` while on linux/macOS `family` contain only `JetBrains Mono` thus same way of collecting weights with `findFontsSync` does not work. Also for some reason `findFontsSync` always returns empty array on windows, even if i use some hardcoded values like `Consolas` or `JetBrains Mono`.

Not sure what would be the best approach in that case, getting all the weights for each font immediately might also result in different behavior on windows since it shows each installed font in separate `family` :disappointed: 

Any suggestions how i can tackle this would be much appreciated.

danilkhromov committed:
fix monospace font filtering

develohpanda commented:
Thanks for that investigation! We all just love the differences between operating systems 😝 I'll spend some time on Windows next week and get back to you!

jackkav commented:
Closed as stale.
